clang/lib/Parse/ParseDeclCXX.cpp

```diff
@@ -3139,6 +3139,19 @@ Parser::DeclGroupPtrTy Parser::ParseCXXClassMemberDeclaration(
     return Actions.BuildDeclaratorGroup(Decls);
   }
 
+  // Befriending a concept is invalid and would already fail if
+  // we did nothing here, but this allows us to issue a more
+  // helpful diagnostic.
+  if (Tok.is(tok::kw_concept)) {
```

Comment:
Could we get a standard reference for a befriending a concept being invalid.

Reply:
> Could we get a standard reference for a befriending a concept being invalid.

Pretty sure the syntax just doesn’t allow for it: Neither a `simple-type-specifier`, nor an `elaborated-type-specifier`, nor a `typename-specifier` can start with `concept`. The grammatical productions for `friend-type-specifier` are already listed in the comment at the top of the function.